src/replication/coordinator_state.cpp

```diff
@@ -0,0 +1,240 @@
+                                                 return RegisterMainReplicaCoordinatorStatus::END_POINT_EXISTS;
+                                               }
+                                               return RegisterMainReplicaCoordinatorStatus::SUCCESS;
+                                             }},
+                 data_);
+
+  if (endpoint_status != RegisterMainReplicaCoordinatorStatus::SUCCESS) {
+    return endpoint_status;
+  }
+
+  auto &registered_main = std::get<CoordinatorData>(data_).registered_main_;
+  registered_main = std::make_unique<CoordinatorClient>(config);
+  return registered_main.get();
+}
+
+auto CoordinatorState::ShowReplicas() const -> std::vector<CoordinatorEntityInfo> {
+  if (!std::holds_alternative<CoordinatorData>(data_)) {
+    MG_ASSERT(false, "Can't call show replicas on data_, as variant holds wrong alternative");
+  }
+  std::vector<CoordinatorEntityInfo> result;
+  const auto &registered_replicas = std::get<CoordinatorData>(data_).registered_replicas_;
+  result.reserve(registered_replicas.size());
+  std::ranges::transform(registered_replicas, std::back_inserter(result), [](const auto &replica) {
+    return CoordinatorEntityInfo{replica.InstanceName(), replica.Endpoint()};
+  });
+  return result;
+}
+
+auto CoordinatorState::ShowMain() const -> std::optional<CoordinatorEntityInfo> {
+  if (!std::holds_alternative<CoordinatorData>(data_)) {
+    MG_ASSERT(false, "Can't call show main on data_, as variant holds wrong alternative");
+  }
+  const auto &registered_main = std::get<CoordinatorData>(data_).registered_main_;
+  if (registered_main) {
+    return CoordinatorEntityInfo{registered_main->InstanceName(), registered_main->Endpoint()};
+  }
+  return std::nullopt;
+}
+
+auto CoordinatorState::PingReplicas() const -> std::unordered_map<std::string_view, bool> {
+  if (!std::holds_alternative<CoordinatorData>(data_)) {
+    MG_ASSERT(false, "Can't call ping replicas on data_, as variant holds wrong alternative");
+  }
+  std::unordered_map<std::string_view, bool> result;
+  const auto &registered_replicas = std::get<CoordinatorData>(data_).registered_replicas_;
+  result.reserve(registered_replicas.size());
+  for (const CoordinatorClient &replica_client : registered_replicas) {
+    result.emplace(replica_client.InstanceName(), replica_client.DoHealthCheck());
+  }
+
+  return result;
+}
+
+auto CoordinatorState::PingMain() const -> std::optional<CoordinatorEntityHealthInfo> {
+  if (!std::holds_alternative<CoordinatorData>(data_)) {
+    MG_ASSERT(false, "Can't call show main on data_, as variant holds wrong alternative");
+  }
+  const auto &registered_main = std::get<CoordinatorData>(data_).registered_main_;
+  if (registered_main) {
+    return CoordinatorEntityHealthInfo{registered_main->InstanceName(), registered_main->DoHealthCheck()};
+  }
+  return std::nullopt;
+}
+
+auto CoordinatorState::DoFailover() -> DoFailoverStatus {
```

Comment:
Is there some safeguard if called from different session?

Reply:
If you mean concurrency issues, they exist yes because we followed replication. What do you mean by different session?